gcc/rust/resolve/rust-ast-resolve-pattern.cc

```diff
@@ -197,6 +171,160 @@ PatternDeclaration::visit (AST::TuplePattern &pattern)
     }
 }
 
+void
+PatternDeclaration::visit (AST::AltPattern &pattern)
+{
+  // push a new set of 'Or' bindings to the stack. Accounts for the
+  // alternatives. e.g. in `p_0 | p_1`, bindings to the same identifier between
+  // p_0 and p_1 shouldn't cause an error.
+  bindings_with_ctx.push_back (
+    PatternBinding (PatternBoundCtx::Or, std::set<Identifier> ()));
+
+  // This is a hack to avoid creating a separate visitor class for the
+  // consistency checks. We empty out the binding_info_map before each iteration
+  // to separate between the alts' binding_maps. And right after the alt
+  // visit...
+  auto tmp_binding_map = binding_info_map;
+  binding_info_map.clear ();
+
+  std::vector<std::map<Identifier, BindingInfo>> alts_binding_maps;
```

Comment:
I think the C++ STL is really poor at dealing with stuff like this because the template arguments get messy, I think you should make the std::map<Identifier, BindingInfo> a typedef or a separate proxy class it means it decouples it a bit if it needs changed.

Reply:
This would have made it easier for me to review the check_consistency function because it was unclear why iterating over the same vector like this made any sense